add blog github part 2 (#483)
bytebase/bytebase.com: content/blog/github-database-cicd-part-2-github-database-gitops.md

---
title: "How to Setup Database CI/CD with GitHub, Part 2: GitHub.com Database GitOps"
author: Ningjing
published_at: 2022/9/6 13:00:00
feature_image: /static/blog/github-database-cicd-part-2-github-database-gitops/howto-github-2.webp
tags: Education
featured: true
description: This second part will guide you through configuring GitHub.com database GitOps with Bytebase. After following these steps, you can trigger database changes by merging sql files into your GitHub repository.
---

This is a series of articles about Database CI/CD with GitHub

- [The Database CI/CD Best Practice with GitHub](/blog/database-cicd-best-practice-with-github)

- How to Setup Database CI/CD with GitHub Part 1-3
  - [Part 1: Enable SQL Review with GitHub Actions](/blog/github-database-cicd-part-1-sql-review-github-actions)
  - Part 2: GitHub.com Database GitOps (this one)
  - Part 3: Put Them Together

---

In the last article [How to Setup Database CI/CD with GitHub, Part 1: Enable SQL Review with GitHub Actions](/blog/github-database-cicd-part-1-sql-review-github-actions), you learned how to enable SQL review GitHub actions and trigger it automatically when there is a pull request in your repo.

> This second part will guide you through configuring GitHub.com database GitOps with Bytebase. After following these steps, you can trigger database changes by merging sql files into your GitHub repository – which means that your database version can always synchronize with your code.

Reviewer: ```suggestion
This second part will guide you through configuring GitHub.com database GitOps with Bytebase. After following these steps, whenever SQL files carrying the database change are merged into your GitHub repository, the database change pipeline will be automatically triggerred – which means that your database version can always synchronize with your code.
```

> This second part will guide you through configuring GitHub.com database GitOps with Bytebase. After following these steps, you can trigger database changes by merging sql files into your GitHub repository – which means that your database version can always synchronize with your code.

Reviewer: ```suggestion
This second part will guide you through configuring GitHub.com database GitOps with Bytebase. After following these steps, whenever SQL files carrying the database change are merged into your GitHub repository, the database change pipeline will be automatically triggered. Thus your Git repository always keeps the source of truth of your database schema and your live database schema always evolves whenever SQL files change in the Git repository.
```

> ## Step 1 - Run Bytebase in Docker with the URL generated by Ngrok

Reviewer: ngrok instead of Ngrok, same in other places

In this tutorial, you’ll run Bytebase locally using Docker.

> [Ngrok](https://ngrok.com/) is a reverse proxy tunnel, and in our case, we need it for a public network address in order to receive webhooks from GitHub.com. Ngrok we used here is for testing purposes. For production use, we recommend using [Caddy](https://caddyserver.com/).

Reviewer: ```suggestion
[Ngrok](https://ngrok.com/) is a reverse proxy tunnel, and in our case, we need it for a public network address in order to receive webhooks from GitHub.com. Ngrok we used here is for demonstration purposes. For production use, we recommend using [Caddy](https://caddyserver.com/).
```

> [Ngrok](https://ngrok.com/) is a reverse proxy tunnel, and in our case, we need it for a public network address in order to receive webhooks from GitHub.com. Ngrok we used here is for testing purposes. For production use, we recommend using [Caddy](https://caddyserver.com/).

Reviewer: I drafted an illustration https://app.excalidraw.com/s/58zvOJAtWmD/3JXxHISNdOM

You can tweak it and attach it here for better explanation

1. Login to [Ngrok Dashboard](https://dashboard.ngrok.com/get-started/setup) and follow its **Get Started** steps to install and configure.
2. Run `ngrok http 8080` and obtain the public URL:

![ngrok](/static/blog/github-database-cicd-part-2-github-database-gitops/ngrok.webp)

3. Run Bytebase in Docker with the following command.
```bash
docker run --init \
--name bytebase-github \
--restart always \
--add-host host.docker.internal:host-gateway \
--publish 8080:8080 \
--volume ~/.bytebase/data:/var/opt/bytebase \
bytebase/bytebase:1.3.0 \
--data /var/opt/bytebase \
--host https://03f1-103-102-7-52.ngrok.io \
--port 8080 \
--frontend-port 80
```
4. Bytebase is running successfully in Docker, and you can visit it via [https://03f1-103-102-7-52.ngrok.io](https://03f1-103-102-7-52.ngrok.io/)

![docker-running](/static/blog/github-database-cicd-part-2-github-database-gitops/docker-running.webp)

> ## Step 2 - Add GitHub.com as a git provider in Bytebase

Reviewer: ```suggestion
## Step 2 - Add GitHub.com as a Git provider in Bytebase
```

1. Open Bytebase, go to **Settings -> Version Control**, choose GitHub.com, and click **Next**.

![bb-vc](/static/blog/github-database-cicd-part-2-github-database-gitops/bb-vc.webp)

2. Copy the **Authorization callback URL**.

![fill-id-secret](/static/blog/github-database-cicd-part-2-github-database-gitops/fill-id-secret.webp)

3. Open GitHub, and go to **Settings -> Developer Settings -> OAuth Apps**. Click **New OAuth App**.
![github-oauth](/static/blog/github-database-cicd-part-2-github-database-gitops/github-oauth.webp)

4. In the new OAuth App, paste the **Authorization callback URL**, then click **Update Application**.

![auth-callback](/static/blog/github-database-cicd-part-2-github-database-gitops/auth-callback.webp)

5. On the same page, you can also find **Client ID** and **Client secrets**.

![client-id-secrets](/static/blog/github-database-cicd-part-2-github-database-gitops/client-id-secrets.webp)

6. Switch back to the Bytebase console, fill Client ID and Client secrets in the form as **Application ID** and **Secret**.

![fill-id-secret](/static/blog/github-database-cicd-part-2-github-database-gitops/fill-id-secret.webp)

> 6. Click **Next**. You will be redirected to the confirmation page. Click **Confirm and add**, and the git provider is successfully added.

Reviewer: ```suggestion
6. Click **Next**. You will be redirected to the confirmation page. Click **Confirm and add**, and the Git provider is successfully added.
```

![transfer-in-db-2](/static/blog/github-database-cicd-part-2-github-database-gitops/transfer-in-db-2.webp)

3. Go to **Version Control** under the project, and choose **GitOps workflow**.

![project-vc](/static/blog/github-database-cicd-part-2-github-database-gitops/project-vc.webp)

![project-vc-2](/static/blog/github-database-cicd-part-2-github-database-gitops/project-vc-2.webp)

4. Create a repository named `database-gitops-sample` in GitHub. Select this repository in Bytebase. Keep the default setting unless you know what you are doing. Pay attention to **Branch** field to make sure it’s the one you want to watch.

![project-repo](/static/blog/github-database-cicd-part-2-github-database-gitops/project-repo.webp)

5. Click **Finish**, and you can see that Version Control is **Enabled**.

![vc-enable](/static/blog/github-database-cicd-part-2-github-database-gitops/vc-enable.webp)

6. Create a **bytebase** folder and **prod**, **test** folders under it in your GitHub repository.

![bytebase-folder](/static/blog/github-database-cicd-part-2-github-database-gitops/bytebase-folder.webp)

> 7. Create a sql script following the name convention `{{ENV_NAME}}/{{DB_NAME}}__{{VERSION}}__{{TYPE}}__{{DESCRIPTION}}.sql`,and here we create `employeeGitHub__202208171630__migrate__add_nickname.sql`.

Reviewer: ```suggestion
7. Create a sql script following the name convention `{{ENV_NAME}}/{{DB_NAME}}__{{VERSION}}__{{TYPE}}__{{DESCRIPTION}}.sql`,and here we create `employeeGitHub__202208171630__migrate__add_nickname.sql` under the `test` directory.
```

Reviewer: Let's further iterate each component

* `test` corresponds to {{ENV_NAME}}
* `employeeGitHub` corresponds to {{DB_NAME}}
* etc...

![create-migration-sql](/static/blog/github-database-cicd-part-2-github-database-gitops/create-migration-sql.webp)

8. Commit and push this script.
9. Go back to the Bytebase Console, and you can see there is a new issue created.

![new-issue](/static/blog/github-database-cicd-part-2-github-database-gitops/new-issue.webp)

> 10. Visit this issue, and click **Approve**. The SQL will execute against Test environment, and it shows **Done**. You may skip this manual approval step by customization.

Reviewer: ```suggestion
10. Visit this issue, and click **Approve**. The SQL will execute against the `employeeGitHub` database in the `Test` environment, and it shows **Done**. You may skip this manual approval step by customization.
```

Reviewer: > You may skip this manual approval step by customization

Include the link to our doc on how to configure approval policy

![issue-waiting](/static/blog/github-database-cicd-part-2-github-database-gitops/issue-waiting.webp)
![issue-done](/static/blog/github-database-cicd-part-2-github-database-gitops/issue-done.webp)

11. Go to database **employeeGitHub** -> table **employee**, and the nickname column is added there.

![table-employee](/static/blog/github-database-cicd-part-2-github-database-gitops/table-employee.webp)

12. Copy the script file and paste it into prod directory, and repeat the process. The schema change will execute on Prod environment.

> ![repeat-product](/static/blog/github-database-cicd-part-2-github-database-gitops/repeat-product.webp)

Reviewer: wrong screenshot

> Congratulations! Now you have 1) Enabled SQL Review GitHub Actions for your GitHub repo, 2) Tried out database GitOps with GitHub. In the next part, we’ll put these two parts together. Stay tuned!

Reviewer: ```suggestion
Congratulations! Now you have 1) Enabled SQL Review GitHub Actions for your GitHub repo, 2) Tried out database GitOps with GitHub. Next in the final part, we’ll put these two parts together. Stay tuned!
```
